File: tools/probe/trace_dispatch.py
#!/usr/bin/env python3
"""Count cipher bytes per kernel by interposing at the link, not by patching text.

Each traced kernel is renamed inside its object and replaced by a shim that bumps
a call counter, adds %rdx (the length argument for every ChaCha entry here) to a
byte counter, and jumps to the original.  Cross-object calls from aead512_core.o
therefore land on the shim; intra-object jumps are unaffected, which is exactly
the granularity we want -- we are asking which kernels the AEAD *dispatches* to.
"""
import subprocess, sys, os
BUILD = sys.argv[1]
OUT = sys.argv[2]
# symbol -> object that defines it
SYMS = {
 'ChaCha20_16x_mac':      'aead_mac.o',
 'ChaCha20_16x_mac1024':  'aead_mac_partial.o',
 'ChaCha20_16x_mac512':   'aead_mac_partial.o',
 'ChaCha20_16x_mac256':   'aead_mac_partial.o',
 'ChaCha20_16x_mac128':   'aead_mac_partial.o',
 'ChaCha20_17x':          'chacha_kernels_core.o',
 'ChaCha20_17x_key':      'chacha_kernels_core.o',
 'ChaCha20_tail_avx512':  'chacha_kernels_core.o',
 'ChaCha20_16x':          'chacha_kernels_core.o',
 'ChaCha20_16x_tiered':   'chacha_kernels_core.o',
 'ChaCha20_32x':          'chacha_kernels_core.o',
 'poly1305_aead_update_fma_avx512':   'poly1305_ifma.o',
 'poly1305_aead_complete_fma_avx512': 'poly1305_ifma.o',
 'poly1305_aead_finish_fma_avx512':   'poly1305_ifma.o',
 'poly1305_pow32':        'poly1305_pow.o',
 'poly1305_pow44':        'poly1305_pow.o',
 'poly1305_gmul44_fast':  'poly1305_pow.o',
}
objs = sorted(set(SYMS.values()))
os.makedirs(OUT, exist_ok=True)
for o in objs:
    args = []
    for s, obj in SYMS.items():
        if obj == o:
            args += ['--redefine-sym', '%s=real_%s' % (s, s)]
    subprocess.run(['objcopy'] + args + [os.path.join(BUILD, o),
                                         os.path.join(OUT, o)], check=True)
asm = ['.text']
for s in SYMS:
    asm += ['.globl %s' % s, '.type %s,@function' % s, '%s:' % s,
            '\tincq\tcalls_%s(%%rip)' % s,
            '\taddq\t%%rdx,bytes_%s(%%rip)' % s,
            '\taddq\t%%rsi,rsi_%s(%%rip)' % s,
            '\tjmp\treal_%s' % s,
            '.size %s,.-%s' % (s, s)]
asm += ['.data']
for s in SYMS:
    asm += ['.globl calls_%s' % s, 'calls_%s: .quad 0' % s,
            '.globl bytes_%s' % s, 'bytes_%s: .quad 0' % s,
            '.globl rsi_%s' % s, 'rsi_%s: .quad 0' % s]
asm += ['.section .note.GNU-stack,"",@progbits']
open(os.path.join(OUT, 'shim.S'), 'w').write('\n'.join(asm) + '\n')
open(os.path.join(OUT, 'syms.txt'), 'w').write('\n'.join(SYMS) + '\n')
print('shimmed %d symbols across %s' % (len(SYMS), ', '.join(objs)))

File: tools/probe/trace_dispatch.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include "../../chacha20poly1305.h"

#define K(n) extern unsigned long calls_##n, bytes_##n, rsi_##n;
K(ChaCha20_16x_mac) K(ChaCha20_16x_mac1024) K(ChaCha20_16x_mac512)
K(ChaCha20_16x_mac256) K(ChaCha20_16x_mac128) K(ChaCha20_17x) K(ChaCha20_17x_key)
K(ChaCha20_tail_avx512) K(ChaCha20_16x) K(ChaCha20_16x_tiered) K(ChaCha20_32x)
K(poly1305_aead_update_fma_avx512) K(poly1305_aead_complete_fma_avx512)
K(poly1305_aead_finish_fma_avx512) K(poly1305_pow32) K(poly1305_pow44) K(poly1305_gmul44_fast)

struct e { const char *nm; unsigned long *c, *b, *s; };
static struct e T[] = {
#define R(n) { #n, &calls_##n, &bytes_##n, &rsi_##n },
 R(ChaCha20_16x_mac) R(ChaCha20_16x_mac1024) R(ChaCha20_16x_mac512)
 R(ChaCha20_16x_mac256) R(ChaCha20_16x_mac128) R(ChaCha20_17x) R(ChaCha20_17x_key)
 R(ChaCha20_tail_avx512) R(ChaCha20_16x) R(ChaCha20_16x_tiered) R(ChaCha20_32x)
 R(poly1305_aead_update_fma_avx512) R(poly1305_aead_complete_fma_avx512)
 R(poly1305_aead_finish_fma_avx512) R(poly1305_pow32) R(poly1305_pow44) R(poly1305_gmul44_fast)
};
#define N (sizeof T / sizeof T[0])

static uint8_t in[70000], out[70100], key[32], nonce[12], ad[13];

int main(void)
{
    size_t L[] = {1280,1420,4096,16384};
    unsigned long c0[N], b0[N], s0[N];
    for (unsigned i = 0; i < sizeof L / sizeof L[0]; i++) {
        size_t ol;
        for (unsigned k = 0; k < N; k++) { c0[k] = *T[k].c; b0[k] = *T[k].b; s0[k] = *T[k].s; }
        chacha20_poly1305_seal(out, &ol, sizeof out, in, L[i], ad, 13, key, nonce);
        printf("payload %zu:\n", L[i]);
        unsigned long tot = 0, big = 0;
        for (unsigned k = 0; k < N; k++) {
            unsigned long c = *T[k].c - c0[k], b = *T[k].b - b0[k], r = *T[k].s - s0[k];
            int poly = T[k].nm[0] == 'p';
            if (!c) continue;
            printf("    %-34s %2lu call%s  len=%lu\n", T[k].nm, c, c == 1 ? " " : "s",
                   poly ? r : b);
            if (!poly) tot += b;
        }
        printf("    cipher bytes %lu\n\n", tot); (void)big;
    }
    return 0;
}
